File: ts/src/extract.ts
/**
 * ExtractCredential helper — pulls the credential from an incoming HTTP
 * request following the shared 5-service pattern: prefer Authorization:
 * Bearer <token>, fall back to the raw "token" header.
 *
 * See design doc §7 and Go SDK's extract.go.
 */

import { OctoAuthError } from './errors.js'
import {
  classifyCredential,
  KindAPIKey,
  KindBot,
  MinAPIKeyLength,
  PrefixApp,
  PrefixBF,
  PrefixUK,
  type Credential,
} from './verifier.js'

export const HeaderAuthorization = 'authorization'
export const HeaderToken = 'token'
export const BearerScheme = 'Bearer'

/**
 * HeadersLike is the minimal request shape extractCredential needs. Both
 * Express's IncomingMessage.headers (Record<string, string | string[] | undefined>)
 * and a plain lowercased map satisfy this.
 */
export interface HeadersLike {
  [name: string]: string | string[] | undefined
}

export interface RequestLike {
  headers: HeadersLike
}

/**
 * extractCredential reads the Authorization: Bearer or "token" header and
 * returns a classified Credential.
 *
 * Throws {@link OctoAuthError} with kind=invalid-credential when no credential
 * is present or classification fails.
 */
export function extractCredential(req: RequestLike): Credential {
  if (req == null || req.headers == null) {
    throw OctoAuthError.invalidCredential('nil request')
  }
  const raw = readCredentialHeader(req.headers)
  if (raw === '') {
    throw OctoAuthError.invalidCredential('missing credential')
  }
  const kind = classifyCredential(raw)
  return { kind, raw }
}

/**
 * readCredentialHeader reads Authorization: Bearer with fallback to "token".
 * Header lookup is case-insensitive; Express lowercases header names but the
 * fallback accommodates arbitrary case for defensive callers.
 */
function readCredentialHeader(h: HeadersLike): string {
  const auth = getHeader(h, HeaderAuthorization)
  if (auth) {
    const token = stripBearer(auth)
    if (token !== '') return token
    // Fall through: an Authorization header without a bearer token is treated
    // the same as no header at all (matches docs-backend behavior).
  }
  const tok = getHeader(h, HeaderToken)
  return tok?.trim() ?? ''
}

/** getHeader looks up name case-insensitively and coerces array values. */
function getHeader(h: HeadersLike, name: string): string | undefined {
  const lower = name.toLowerCase()
  // Fast path: already lowercased (Express, Hono, fetch Headers via .toJSON).
  const direct = h[lower]
  if (direct !== undefined) return coerceHeader(direct)
  // Fall back: scan keys case-insensitively.
  for (const k of Object.keys(h)) {
    if (k.toLowerCase() === lower) {
      return coerceHeader(h[k])
    }
  }
  return undefined
}

function coerceHeader(v: string | string[] | undefined): string | undefined {
  if (v === undefined) return undefined
  if (Array.isArray(v)) return v[0]
  return v
}

/**
 * stripBearer returns the token part of "Bearer <token>" (scheme match is
 * case-insensitive per RFC 7235). Returns "" if value is not a bearer credential.
 */
function stripBearer(v: string): string {
  const trimmed = v.trim()
  if (trimmed.length <= BearerScheme.length) return ''
  if (trimmed.slice(0, BearerScheme.length).toLowerCase() !== BearerScheme.toLowerCase()) {
    return ''
  }
  const rest = trimmed.slice(BearerScheme.length)
  if (rest.length === 0 || (rest[0] !== ' ' && rest[0] !== '\t')) return ''
  return rest.trim()
}

/**
 * validateAPIKey checks that token has the "uk_" prefix and meets the length
 * requirement. Useful for producers minting their own keys.
 */
export function validateAPIKey(token: string): void {
  if (!token.startsWith(PrefixUK)) {
    throw OctoAuthError.invalidCredential('api key missing uk_ prefix', {
      verifier: KindAPIKey,
    })
  }
  if (token.length < MinAPIKeyLength) {
    throw OctoAuthError.invalidCredential('api key too short', {
      verifier: KindAPIKey,
    })
  }
}

/** validateBotToken checks that token has one of the bot-realm prefixes. */
export function validateBotToken(token: string): void {
  if (token.startsWith(PrefixBF) || token.startsWith(PrefixApp)) {
    if (token.length <= PrefixBF.length) {
      throw OctoAuthError.invalidCredential('bot token has prefix only', {
        verifier: KindBot,
      })
    }
    return
  }
  throw OctoAuthError.invalidCredential('bot token missing bf_/app_ prefix', {
    verifier: KindBot,
  })
}

File: go/extract.go
package octoauth

import (
	"net/http"
	"strings"
)

// Header names and credential-prefix constants used across the SDK. Callers
// referring to these names in middleware or helpers should import the
// constants rather than hard-coding strings so a rename in the future stays
// in one place.
const (
	// HeaderAuthorization is the RFC 7235 authorization header.
	HeaderAuthorization = "Authorization"
	// HeaderToken is the octo-server-native token header used by some
	// clients (mirrors what octo-server itself accepts as a fallback).
	HeaderToken = "token"
	// BearerScheme is the (case-insensitive) prefix for bearer tokens.
	BearerScheme = "Bearer"

	// PrefixUK marks a userKeyVerifier credential.
	PrefixUK = "uk_"
	// PrefixBF marks a botTokenVerifier credential.
	PrefixBF = "bf_"
	// PrefixApp is reserved for the future App Bot token realm; v1 forwards
	// it to the bot verifier so a future rollout stays additive.
	PrefixApp = "app_"

	// MinAPIKeyLength is the minimum total length (including the "uk_"
	// prefix) that a user API key must have. Matches octo-fleet's
	// `len >= 35` guard and the design-doc §3 dispatch rule.
	MinAPIKeyLength = 35
)

// ExtractCredential pulls the credential from an HTTP request following the
// pattern shared by the 5 surveyed services: prefer Authorization: Bearer
// <token>, fall back to the raw "token" header. The returned [Credential] has
// its Kind populated by prefix inspection (session / bot / apikey).
//
// Returns [*Error] with Kind ErrKindInvalidCredential when no credential is
// present, when an Authorization header uses a non-bearer scheme with an
// empty token, or when a "uk_"-prefixed credential is shorter than
// [MinAPIKeyLength].
func ExtractCredential(r *http.Request) (Credential, error) {
	if r == nil {
		return Credential{}, &Error{
			Kind:    ErrKindInvalidCredential,
			Message: "nil request",
		}
	}
	raw := readCredentialHeader(r.Header)
	if raw == "" {
		return Credential{}, &Error{
			Kind:    ErrKindInvalidCredential,
			Message: "missing credential",
		}
	}
	kind, err := classifyCredential(raw)
	if err != nil {
		return Credential{}, err
	}
	return Credential{Kind: kind, Raw: raw}, nil
}

// readCredentialHeader reads an Authorization: Bearer token (case-insensitive
// scheme) with fallback to the raw "token" header. Returns the trimmed token
// or "" if neither header carries one.
func readCredentialHeader(h http.Header) string {
	if auth := h.Get(HeaderAuthorization); auth != "" {
		if token := stripBearer(auth); token != "" {
			return token
		}
		// An Authorization header without a bearer token is treated the
		// same as no header at all — fall through to the "token" header
		// rather than erroring out, matching docs-backend's extractOctoToken
		// behavior.
	}
	if tok := strings.TrimSpace(h.Get(HeaderToken)); tok != "" {
		return tok
	}
	return ""
}

// stripBearer returns the token part of "Bearer <token>" (scheme match is
// case-insensitive per RFC 7235). Returns "" if the value is not a bearer
// credential.
func stripBearer(v string) string {
	trimmed := strings.TrimSpace(v)
	if len(trimmed) <= len(BearerScheme) {
		return ""
	}
	if !strings.EqualFold(trimmed[:len(BearerScheme)], BearerScheme) {
		return ""
	}
	rest := trimmed[len(BearerScheme):]
	// Scheme must be followed by whitespace per RFC 7235; be lenient and
	// accept any whitespace run.
	if len(rest) == 0 || (rest[0] != ' ' && rest[0] != '\t') {
		return ""
	}
	return strings.TrimSpace(rest)
}

// ValidateAPIKey checks that token has the "uk_" prefix and meets the length
// requirement. Returns nil on success or [*Error] Kind ErrKindInvalidCredential
// on failure. Useful for producers who mint their own keys and want to catch
// obvious mistakes before shipping them.
func ValidateAPIKey(token string) error {
	if !strings.HasPrefix(token, PrefixUK) {
		return &Error{
			Kind:     ErrKindInvalidCredential,
			Message:  "api key missing uk_ prefix",
			Verifier: KindAPIKey,
		}
	}
	if len(token) < MinAPIKeyLength {
		return &Error{
			Kind:     ErrKindInvalidCredential,
			Message:  "api key too short",
			Verifier: KindAPIKey,
		}
	}
	return nil
}

// ValidateBotToken checks that token has one of the bot-realm prefixes ("bf_"
// or the reserved "app_"). Returns nil on success or [*Error] on failure.
func ValidateBotToken(token string) error {
	if strings.HasPrefix(token, PrefixBF) || strings.HasPrefix(token, PrefixApp) {
		if len(token) <= len(PrefixBF) {
			return &Error{
				Kind:     ErrKindInvalidCredential,
				Message:  "bot token has prefix only",
				Verifier: KindBot,
			}
		}
		return nil
	}
	return &Error{
		Kind:     ErrKindInvalidCredential,
		Message:  "bot token missing bf_/app_ prefix",
		Verifier: KindBot,
	}
}
